src/style-spec/function/types.js

```diff
@@ -54,6 +54,44 @@ function toString(type: Type): string {
     }
 }
 
+const valueMemberTypes = [
+    NullType,
+    NumberType,
+    StringType,
+    BooleanType,
+    ColorType,
+    ObjectType,
+    array(ValueType)
+];
+
+/**
+ * Returns null if `t` is a subtype of `expected`; otherwise returns an
+ * error message.
+ * * @private
+ */
+function checkSubtype(expected: Type, t: Type): ?string {
+    if (t.kind === 'Error') {
+        // Error is a subtype of every type
+        return null;
+    } else if (expected.kind === 'Array') {
+        if (t.kind === 'Array' &&
+            !checkSubtype(expected.itemType, t.itemType) &&
+            (typeof expected.N !== 'number' || expected.N === t.N)) {
+            return null;
+        }
+    } else if (expected.kind === t.kind) {
+        return null;
+    } else if (expected.kind === 'Value') {
+        for (const memberType of valueMemberTypes) {
+            if (!checkSubtype(memberType, t)) {
+                return null;
+            }
+        }
```

Comment:
Hm. I'm realizing now: according to this logic (together with array subtyping), `Value` is effectively the `top` type... should `Color` really be a member type of `Value`?

Reply:
Let's ticket this separately.

Reply:
Ticketed #5296